File: BingoCTF_2020/guestbook/worker/worker.js
const fs = require('fs')
const md5 = require('md5');

const puppeteer = require('puppeteer');
const Redis = require('ioredis');
const connection = new Redis(6379, 'redis');

const admin_username = "admin";
const admin_password = "__4DM1N_P4SSW0RD__";

const browser_option = {
    executablePath: 'google-chrome-unstable',
    headless: true,
    args: [
        '--no-sandbox',
        '--disable-background-networking',
        '--disable-default-apps',
        '--disable-extensions',
        '--disable-gpu',
        '--disable-sync',
        '--disable-translate',
        '--hide-scrollbars',
        '--metrics-recording-only',
        '--mute-audio',
        '--no-first-run',
        '--safebrowsing-disable-auto-update',
    ],
};

let browser = undefined;

const crawl = async (url) => {
    console.log(`[+] Query! (${url})`);
    const page = await browser.newPage();
    try {
        await page.goto(url, {
            waitUntil: 'networkidle0',
            timeout: 5 * 1000,
        });
        // await page.on("console", msg => { console.log(msg.text()); });
        // await page.waitFor('#answer');
        // await page.type('#answer', 'ㄴ Thanks for messaging! I am really appreciated! :)');
        // await page.waitFor('#answer-button');
        // await Promise.all([
        //    page.$eval('#answer-button', elem => elem.click()),
        //    page.waitForNavigation()
        //]);
        await page.waitForNavigation({timeout: 5000});
    } catch (err){
        console.log(err);
    }
    await page.close();
    console.log(`[+] Done! (${url})`)
};

const init = async () => {
    const browser = await puppeteer.launch(browser_option);
    const page = await browser.newPage();
    console.log(`[+] Setting up...`);
    try {
        await page.goto('http://public/');
        await page.waitFor('#username');
        await page.type('#username', admin_username);
        await page.waitFor('#password');
        await page.type('#password', admin_password);
        await page.waitFor('#login-button');
        await Promise.all([
            page.$eval('#login-button', elem => elem.click()),
            page.waitForNavigation()
        ]);
        const body = await page.evaluate(() => document.body.innerHTML);
        if (!body.includes('Answer')){
            throw Error(`Login failed at ${page.url()}.`);
        }
        console.log(`[+] Setup done!`);
    } catch (err) {
        console.log(`[-] Error while setting up :(`);
        console.log(err);
        const body = await page.evaluate(() => document.body.innerHTML);
        console.log(`body: ${body}`);
    }
    try{
        await page.close();
    } catch (err) {
        console.log(err);
    }
    return browser;
};

function handle(){
    console.log("[+] handle");
    connection.blpop("query", 0, async function(err, message) {
        if (browser === undefined) browser = await init();
        await crawl("http://public/index.php?id=" + message[1]);
        setTimeout(handle, 10); // handle next
    });
}
handle();

File: 2021_ACSC/baby-developer/mobile-viewer/app/worker/worker.py
#!/usr/bin/python3 -u
#-*- coding: utf-8 -*-
# Developer: stypr (https://harold.kim/)

import asyncio
import time
import redis
import requests
from pyppeteer import launch

redis = redis.Redis('redis')
browser = None
browser_option = {
    'executablePath': '/usr/bin/google-chrome-stable',
    'options': {'args': [
        '--no-sandbox',
        '--disable-default-apps',
        '--disable-extensions',
        '--disable-gpu',
        '--disable-sync',
        '--disable-translate',
        '--hide-scrollbars',
        '--metrics-recording-only',
        '--mute-audio',
        '--no-first-run',
        '--safebrowsing-disable-auto-update',
        '--disable-dev-shm-usage', # need review
        '--user-agent=Mozilla/5.0 (iPhone; CPU iPhone OS 12_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148',
    ]},
    'handleSIGINT': False,
    'handleSIGTERM': False,
    'handleSIGHUP': False,
    'headless': True
}

async def init_browser():
    global browser
    _browser = await launch(**browser_option)
    page = await _browser.newPage()
    try:
        await page.goto('http://example.com/')
    except Exception as e:
        print("[!] Error during browser initialization: " + e)
    finally:
        await page.close()
    print("[.] Browser is now loaded.")
    return _browser

async def render(uuid, url):
    result = f"../static/output/{uuid}.jpg"
    timeout = 3
    try:
        print("[.] Begin crawling...")
        page = await browser.newPage()
        await page.goto(url, {
            'timeout': timeout * 1000,
            'waitUntil' : 'networkidle0'
        })
        print("[.] Mobile Rendering complete..")
        # For super extreme browser rendering
        # Credits to tyage for reporting an excellent bug :)
        await page.setViewport({
            'width': 16, # 640//2,
            'height': 16, #1136//2})
        })
        await page.screenshot({'path': result, "type": "jpeg"})
        print(f"[.] Done saving the file. ({uuid})")

    except Exception as e:
        print(f"[!!] {e}")
    finally:
        await page.close()

async def main(loop):
    global browser
    while True:
        try:
            print("[+] Handler has begun..")
            data = redis.blpop("query")[1].decode().split("/")
            if data:
                uuid, url = data[0], '/'.join(data[1:])
                print(f"[*] Got a new request: {uuid} => {url}")
                if not browser:
                    browser = await init_browser()
                _task = await asyncio.wait_for(render(uuid, url), timeout=5)

        except Exception as e:
            print(f"[!] {e}")

        finally:
            await asyncio.sleep(0.01)

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main(loop))
